### PR TITLE
Default values on initialization

### LINKED ISSUE #66: Default proc if empty
Is there a way to set a default value for a list using a proc?  I'm thinking something like:

```ruby
class User < ApplicationRecord
  kredis_list :role_ids, default: ->(user) { user.roles.pluck(:id) }
end
```

Then if there's a nil value (either because it was never created or expired at some point), it recreates itself with the latest data.

I can accomplish the same thing with:

```ruby
class User < ApplicationRecord
  kredis_list :role_ids_cache, default: ->(user) { user.roles.pluck(:id) }

  def role_ids
    self.role_ids_cache.append(self.roles.pluck(:id)) unless self.role_ids_cache.elements.empty?
    self.role_ids_cache
  end
end
```

... but I prefer dealing with one method/variable vs two.

### PR DESCRIPTION
**Pull request by lewispb**
I built upon what was started in https://github.com/rails/kredis/pull/89 but attempted a simpler implementation, only setting default values upon type initialization.

Closes #66 and #89

Almost all the same tests pass, with one or two minor differences.

**Todo**
- [x] Update readme
- [x] Code style (see https://github.com/rails/kredis/pull/120)

### CODE REVIEW

File: test/types/enum_test.rb [hunk 1]
```diff
@@ -37,9 +49,6 @@ class EnumTest < ActiveSupport::TestCase
   end
 
   test "exists?" do
-    assert_not @enum.exists?
-
-    @enum.value = "one"
```
Comment (lewispb) [inline test/types/enum_test.rb:42]: An enum always has a default and so always now exists.

**Comment by dhh** [comment]
I like this a lot. Nice simplification 👍

**Review by jeremy — COMMENTED** [review]
Enum is a bit of an oddball re. defaults out of the enumerated set and `nil` defaults in particular. From HEY:
```
kredis_enum :status, values: %w[ waiting confirmed ], default: nil, key: …
```
Defaults out of the enumerated set should probably be considered a bug, with a special case for `default: nil` behaving how an `optional: true` argument would.

File: lib/kredis/default_values.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+      def set_default
+        raise NotImplementedError, "Kredis type #{self.class} needs to define #set_default"
+      end
+  end
+
+  def initialize(...)
+    super
+
+    if default.present? && !exists?
+      set_default
```
[the remark this answers is not in the document]
Comment (lewispb) [reply lib/kredis/default_values.rb]: Great point. Not possible with MULTI, can't check-and-set. Implemented with WATCH, and tested.

File: lib/kredis/attributes.rb [hunk 3]
```diff
@@ -90,6 +90,7 @@ def kredis_connection_with(method, name, key, **options)
           if instance_variable_defined?(ivar_symbol)
             instance_variable_get(ivar_symbol)
           else
+            options[:default] = kredis_default_evaluated(options[:default]) if options[:default]
             new_type = Kredis.send(type, kredis_key_evaluated(key) || kredis_key_for_attribute(name), **options)
```
[the remark this answers is not in the document]
Comment (lewispb) [reply lib/kredis/attributes.rb:94]: cycle, slot and slots types don't support a default kwarg, so can't blindly send that kwarg on to all types if it's not supplied.

File: lib/kredis/types/enum.rb [hunk 4]
```diff
@@ -33,4 +38,8 @@ def define_predicates_for_values
         define_singleton_method("#{defined_value}!") { self.value = defined_value }
       end
     end
+
+    def set_default
+      self.value = default
+    end
```
[the remark this answers is not in the document]
Comment (lewispb) [reply lib/kredis/types/enum.rb:44]: Fixed!

**Comment by lewispb** [comment]
> Defaults out of the enumerated set should probably be considered a bug, with a special case for nil

Thanks @jeremy, agree, and updated to raise an error in that case.

**Review by jeremy — APPROVED** [review]

File: README.md [hunk 5]
```diff
@@ -197,6 +189,29 @@ person.morning.value = "blue"                        # => SET people:5:morning
...
+  kredis_list :favorite_colors, default: [ "red", "green", "blue" ]
+end
+```
+
+There's a performance overhead to consider though. When you first read or write an attribute in a model, Kredis will
+check if the underlying Redis key exists, while watching for concurrent changes, and if it does not,
+write the specified default value.
+
+This means that using default values in a typical Rails app additional Redis calls (WATCH, EXISTS, UNWATCH) will be
+executed for each Kredis attribute with a default value read or written during a request.
```
Comment (tleish) [inline README.md:211]: Does this occur every request, or only when a value does not exist?

Comment (lewispb) [reply README.md:211]: In the context of a Rails web request, as the Kredis object is initialized, the key is checked with WATCH, EXISTS, UNWATCH once, but then subsequent operations within the request do not incur any Redis command overhead. If no default value is specified, no additional Redis command overhead is incurred at all.

Comment (lewispb) [reply README.md:211]: ### Scenario 1: Default NOT defined
(no additional overhead)

### Scenario 2: Default defined, value NOT exists
WATCH, EXISTS, write default (depending on type), UNWATCH

### Scenario 3: Default defined, value exists
WATCH, EXISTS, UNWATCH

Comment (tleish) [reply README.md:211]: It seems a more efficient approach would be to only add extra overhead of default is defined AND value doesn't exist. In other words:

1. Get Value
2. Value doesn't exist
3. Execute Default Logic

This avoids extra overhead if the value has been set already.

Comment (lewispb) [reply README.md:211]: I agree @tleish. That's exactly scenario 3, but with the added WATCH / UNWATCH.

To illustrate the problem that the WATCH / UNWATCH solves check this test case:

https://github.com/rails/kredis/blob/b1a818ad3e70a23154ed90105a6f1d6f060cbbd1/test/types/counter_test.rb#L113-L121

Without the WATCH / UNWATCH, the value existing check is prone to a race condition.

**Comment by nzwsch** [comment]
I noticed that the default value cannot be set in kredis_integer in the current release version 1.5.0.
```
Failure/Error: kredis_integer :foo, default: 0

ArgumentError:
  unknown keyword: :default
```
When will kredis be released with this PR?

**Comment by dhh** [comment]
1.6.0 released 👍

**Pull request merged**